lib/service/dataLink.js

```diff
@@ -0,0 +1,214 @@
+/**
+ * Copyright (c) 2016 Kinvey Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+'use strict';
+
+const _ = require('lodash');
+const kinveyCompletionHandler = require('./kinveyCompletionHandler');
+const kinveyErrors = require('kinvey-datalink-errors');
+const notImplementedHandler = require('./notImplementedHandler');
+
+let serviceObject = '';
+
+let registeredServiceObjects = {};
+const dataRegistrationOperators = [
+  'onInsert',
+  'onDeleteById',
+  'onDeleteAll',
+  'onDeleteByQuery',
+  'onUpdate',
+  'onGetById',
+  'onGetAll',
+  'onGetByQuery',
+  'onGetCount',
+  'onGetCountWithQuery'
+];
+
+class ServiceObject {
+  constructor(serviceObjectName) {
+    this.eventMap = {};
+    this.serviceObjectName = serviceObjectName;
+  }
+
+  register(dataOp, functionToExecute) {
+    if ((typeof dataOp === 'undefined' && dataOp === null) ||
+      dataRegistrationOperators.indexOf(dataOp) < 0) {
+      throw new Error('Operation not permitted');
+    }
+
+    this.eventMap[dataOp] = functionToExecute;
+  }
+
+  unregister(dataOp) {
+    if ((typeof dataOp === 'undefined' && dataOp === null) ||
+      dataRegistrationOperators.indexOf(dataOp) < 0) {
+      throw new Error('Operation not permitted');
+    }
+
+    delete this.eventMap[dataOp];
+  }
+
+  onInsert(functionToExecute) {
+    this.register('onInsert', functionToExecute);
+  }
+
+  onDeleteById(functionToExecute) {
+    this.register('onDeleteById', functionToExecute);
+  }
+
+  onDeleteAll(functionToExecute) {
+    this.register('onDeleteAll', functionToExecute);
+  }
+
+  onDeleteByQuery(functionToExecute) {
+    this.register('onDeleteByQuery', functionToExecute);
+  }
+
+  onUpdate(functionToExecute) {
+    this.register('onUpdate', functionToExecute);
+  }
+
+  onGetById(functionToExecute) {
+    this.register('onGetById', functionToExecute);
+  }
+
+  onGetAll(functionToExecute) {
+    this.register('onGetAll', functionToExecute);
+  }
+
+  onGetByQuery(functionToExecute) {
+    this.register('onGetByQuery', functionToExecute);
+  }
+
+  onGetCount(functionToExecute) {
+    this.register('onGetCount', functionToExecute);
+  }
+
+  onGetCountWithQuery(functionToExecute) {
+    this.register('onGetCountWithQuery', functionToExecute);
+  }
+
+  removeHandler(handler) {
+    this.unregister(handler);
+  }
+
+  resolve(dataOp) {
+    if (!this.eventMap[dataOp]) {
+      return notImplementedHandler;
```

Comment:
Unless I'm missing something, there's something very wonky going on here... notImplementdHandler is defined as
```
const notImplementedHandler = (request, complete) => {
  complete('These methods are not implemented').notImplemented().done();
}
```

But this `resolve` function is just passed a callback in at least one place. I'm not sure what you expect `complete` to be, but it looks like it's an implicit dependency two-levels deep (it has to be a function that accepts a string and returns an object containing a `notImplemented` function that returns an object containing a `done` function)?